restli-common/src/main/java/com/linkedin/restli/common/validation/RestLiDataValidator.java

```diff
@@ -274,9 +278,14 @@ else if (annotationName.equals(CreateOnly.class.getAnnotation(RestSpecAnnotation
     _validatorClassMap = Collections.unmodifiableMap(validatorClassMap);
   }
 
-  private class DataValidator extends DataSchemaAnnotationValidator
+  /**
+   * Validates input data and patches using a given resource's {@link ReadOnly} and {@link CreateOnly} annotations.
+   * Since it's an extension of {@link DataSchemaAnnotationValidator}, it also validates the data using whatever custom
+   * validators are defined in the schema.
+   */
+  private class RestLiAnnotationDataValidator extends DataSchemaAnnotationValidator
```

Comment:
I totally agree the naming needs to be enhanced. There are just too many validators IMO.

I was always curious about this naming convention. Before we named validators like "DataSchemaAnnotationValidator" or "DataValidator", now later a few new validators are now with a "RestLi"prefix, does that carry special meaning?

Reply:
Good question. My reasoning is:
- The `"RestLi"` prefix implies that the context is the Rest.li request/response validation, not just plain Pegasus data.
- `"Annotation"` because the validator triggers the custom schema validators as well as the "Rest.li annotations" (see next bullet point).
- Together, `"RestLiAnnotation"` implies that the validator uses "Rest.li annotations" (i.e. `ReadOnly`/`CreateOnly`). These are annotations that are specific to the Rest.li REST framework, not just the Pegasus data layer.

Naming is often very confusing or ambiguous in Rest.li, so I'm trying to make it more clear.

Reply:
(1) Maybe out of scope but I think "DataSchemaAnnotationValidator" is itself ambiguous, which made RestLiAnnotationDataValidator ambiguous. Do you see room to improve?  I think it is still a "**DataValidator**", and instantiated from "schemaProperty", so IMO it should be something like "SchemaCustom**DataValidator**"
(2) For "RestLiAnnotationDataValidator" according to what you said, I recommend "RestLiResourceAnnotationSchemaCustomDataValidator".

Reply:
1. I agree 100%, I was also confused by this. The `DataSchemaAnnotationValidator` is itself a data validator, not a schema validator (or a schema annotation validator). It doesn't sound smooth, but the most accurate name would be `CustomValidatorDataValidator` or `SchemaValidatorDataValidator` (the usage of "validator" for two different things is annoying). The name could be improved, however since this class is public changing the name has compatibility concerns. We can change it, but it's a little too much for this change IMO.
2. That suggested name is a mouthful, how about `ResourceAnnotationDataValidator` or `RestLiResourceAnnotationDataValidator`? If we get around to changing the name of the other class, we can change this one too to be consistent with the other (it's private so changing it again is no big deal).

Reply:
For (2) RestLiResourceAnnotationDataValidator sounds good enough, can we use that?

For (1), yes, I know. Also don't know if we can really change that one day (since it is Public), lol

Reply:
Yeah, I'll update the PR to use that.
> yes, I know. Also don't know if we can really change that one day (since it is Public)

Actually, I think we can maintain some compatibility by renaming the class, then preserving the old class as a deprecated subclass. Eventually, we would delete the old class.